### PR TITLE
[Bugfix] Add scf ForOp support and nested loop search in dataflow and pipeline

### PR DESCRIPTION
**Pull request by AdrianLiu00**
## Description ##
1. Add scf ForOp support for dataflow() and pipeline() in customize.py;
2. Add nested loop search support for dataflow() in customize.py

### Problems ###
1. Absence of scf ForOp support in some functions.
2. Fail to search the inner loop when it's deeply nested when executing dataflow().

### Proposed Solutions ###
Modify the recursive searching in nested loop.

### Examples ###
Provided in the test_schedule_stream.py.

## Checklist ##

- [√] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [√] Changes are complete (i.e. I finished coding on this PR)
- [√] All changes have test coverage (It would be better to provide ~2 different test cases to test the robustness of your code)
- [ ] Code is well-documented

### CODE REVIEW

File: allo/customize.py [hunk 1]
```diff
@@ -435,24 +433,25 @@ def dataflow(self, axis):
         band_name = band_name.split(":")[1]
         cnt = 0
 
-        # TODO: Fix deep nested
-        def DFS(op):
+        def locateLoop(op):
```
Comment (chhzh123) [inline allo/customize.py]: Follow [Python naming convention](https://peps.python.org/pep-0008/): `locate_loop`

File: tests/test_schedule_stream.py [hunk 2]
```diff
@@ -120,5 +120,46 @@ def top(A: T[10, 20]) -> T[10, 20]:
...
+        for i1 in range(times_out, name="out1"):
+            for i2 in range(A, name="out2"):
+                for j in range(B, name="target"):
+                    for k in range(N, name="in1"):
+                        C[k] = C[k] + 1
+                    for k in range(N, name="in2"):
+                        C[k] = C[k] - 1
+
+    s1 = allo.customize(case1, instantiate=[8])
+    target_loop = s1.get_loops(s1.top_func_name)["out1"]["j"]
```
Comment (chhzh123) [inline tests/test_schedule_stream.py:158]: Why "j" here would work? I'm still debating whether the "j" loop should be exposed at the outer loop level or should access it by `["out1"]["i2"]["j"]`

**Review by chhzh123 — COMMENTED** [review]
Thanks for the PR! I see you didn't use the `band_name`. I'm wondering if it will be a problem if there are multiple for loops using the same variable name. We need some mechanism to distinguish between loops.

**Comment by AdrianLiu00** [comment]
Yeah, it's all about the way that customize() get access to the inner loop. Since the nested loops are now all exposed to the outermost loop, so I just use `band_name` in the condition to start the recursive locateLoop() instead inside the traversal procedure.
```
        for op in func.entry_block.operations:
            if isinstance(op, (scf_d.ForOp, affine_d.AffineForOp)):
                if (
                    "op_name" in op.attributes
                    and op.attributes["op_name"].value == band_name
                ):
                    locateLoop(op)
```

In my opinion, it's more intuitive for user to access inner loop by `["out1"]["i2"]["j"]`. And it will make locate the specific loop esaier than the traversal procedure now. But this modification could be possibly thorough, because all previously relevant part should be modified.

We can discuss further about how to improve it. Also, the problem to identify the loops using the same variable name is supposed to address as well maybe.

**Comment by chhzh123** [comment]
Ah sorry, I didn't notice you used `band_name`. You can fix the naming issue and go ahead. For the loop access, we can open a new issue to discuss.

**Pull request merged**